### PR TITLE
Synchronized

### PR DESCRIPTION
**Pull request by orsenkucher**
## Status
**KINDA READY**

## Breaking Changes
NO, but storage is no longer singleton 

## Description
Fixing file corruption employing `synchronized` package.
Heavy-write test passes.
Made some adjustments to run tests from single `main`.

Singleton makes things hard to test, so removing it.
## Todos
- [x] Tests

## Impact to Remaining Code Base
This PR will affect:

* pubspec.yaml 
* hydrated_bloc_storage.dart 
* all tests

### CODE REVIEW

File: test/hydrated_bloc_delegate_test.dart [hunk 1]
```diff
@@ -1,144 +0,0 @@
-import 'dart:convert';
```
Comment (felangel) [inline test/hydrated_bloc_delegate_test.dart:1]: why did we delete these tests?

File: test/hydrated_bloc_storage_test.dart [hunk 2]
```diff
@@ -1,195 +0,0 @@
-import 'dart:convert';
```
Comment (felangel) [inline test/hydrated_bloc_storage_test.dart:1]: why did we delete these tests?

File: lib/src/hydrated_bloc_storage.dart [hunk 3]
```diff
@@ -26,31 +27,36 @@ class HydratedBlocStorage implements HydratedStorage {
...
   /// `storageDirectory` can optionally be provided.
   /// By default, `getTemporaryDirectory` is used.
   static Future<HydratedBlocStorage> getInstance({
     Directory storageDirectory,
-  }) async {
-    if (_instance != null) {
-      return _instance;
-    }
+  }) {
+    return lock.synchronized(() async {
```
Comment (felangel) [inline lib/src/hydrated_bloc_storage.dart]: 💯

File: lib/src/hydrated_bloc_storage.dart [hunk 4]
```diff
@@ -26,31 +27,36 @@ class HydratedBlocStorage implements HydratedStorage {
   final Map<String, dynamic> _storage;
   final File _file;
 
+  /// `Lock` is used to synchronize access to IO
+  static final Lock lock = Lock();
```
Comment (felangel) [inline lib/src/hydrated_bloc_storage.dart]: can we make this lock private?

**Comment by felangel** [comment]
@orsenkucher what's the status of this PR? Do you feel it's ready to merge?

**Comment by orsenkucher** [comment]
Yes, sure!

File: lib/src/hydrated_bloc_storage.dart (hunk 4 above)
Comment (orsenkucher) [reply lib/src/hydrated_bloc_storage.dart]: In next PR we did

Comment (orsenkucher) [reply lib/src/hydrated_bloc_storage.dart]: It was need for testing, before making storage non-singleton

**Comment by felangel** [comment]
Can we try to merge https://github.com/felangel/hydrated_bloc/pull/39 into this branch first and then I can do a complete review of all the changes before merging into v3.1.0?

**Comment by orsenkucher** [comment]
@felangel and the next PR is supposed to fix problems of this PR by removing singleton

**Review by felangel — APPROVED** [review]
LGTM 👍 
Thanks so much for taking the time to make these improvements!

**Pull request merged**